rcl_executor/src/let_executor.c

```diff
@@ -0,0 +1,629 @@
+      }
+      break;
+
+    default:
+      RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Error:wait_set unknwon handle type: %d",
+        executor->handles[i].type);
+      return RCL_RET_ERROR;
+  }    // switch-case
+  return rc;
+}
+
+/***
+ * operates on executor->handles[i] object
+ * - calls every callback of each object depending on its type
+ */
+static
+rcl_ret_t
+_rcle_execute(rcle_let_executor_t * executor, rcl_wait_set_t * wait_set, size_t i)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(executor, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT);
+  rcl_ret_t rc = RCL_RET_OK;
+
+  bool invoke_callback = false;
+
+  // determine, if callback shall be called
+  if (executor->handles[i].invocation == ON_NEW_DATA &&
+    executor->handles[i].data_available == true)
+  {
+    invoke_callback = true;
+  }
+
+  if (executor->handles[i].invocation == ALWAYS) {
+    invoke_callback = true;
+  }
+
+  // printf("execute: handles[%d] :  %d\n", i, invoke_callback);  // debug(jst3si)
+  // execute callback
+  if (invoke_callback) {
+    switch (executor->handles[i].type) {
+      case SUBSCRIPTION:
+        executor->handles[i].callback(executor->handles[i].data);
+        break;
+
+      case TIMER:
+        rc = rcl_timer_call(executor->handles[i].timer);
+        if (rc != RCL_RET_OK) {
+          PRINT_RCL_ERROR(rcle_execute, rcl_timer_call);
+          return rc;
+        }
+        break;
+
+      default:
+        RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Execute callback: unknwon handle type: %d",
+          executor->handles[i].type);
+        return RCL_RET_ERROR;
+    }    // switch-case
+  }
+
+  return rc;
+}
+
+static
+rcl_ret_t
+_rcle_let_scheduling(rcle_let_executor_t * executor, rcl_wait_set_t * wait_set)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(executor, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t rc = RCL_RET_OK;
+
+  // logical execution time
+  // 1. read all input
+  // 2. process
+  // 3. write data (*) data is written not at the end of all callbacks, but it will not be
+  //    processed by the callbacks 'in this round' because all input data is read in the
+  //    beginning and the incoming messages were copied.
+
+  // step 1:
+  // take available input data from DDS queue by calling rcl_take()
+  // complexity: O(n) where n denotes the number of handles
+  for (size_t i = 0; (i < executor->max_handles && executor->handles[i].initialized); i++) {
+    rc = _rcle_read_input_data(executor, wait_set, i);
+    if ((rc != RCL_RET_OK) && (rc != RCL_RET_SUBSCRIPTION_TAKE_FAILED)) {
+      return rc;
+    }
+  }  // for-loop
+
+
+  // step 2/ step 3
+  // execute the callbacks in the order of the elements in the array 'executor->handles'
+  // complexity: O(n) where n denotes the number of handles
+  for (size_t i = 0; (i < executor->max_handles && executor->handles[i].initialized); i++) {
+    rc = _rcle_execute(executor, wait_set, i);
```

Comment:
I'll add to this that taking multiple samples of data from the middleware before they're ready, is especially bad (though that doesn't seem to be happening in `_rcle_read_input_data`/`_rcle_let_scheduling` at the moment), since while executing other things those items might have been replaced or removed by the middleware.

Consider that you have a subscription with depth 1 and keep last and you are using lifespan. Perhaps it is ready to be taken from (has 1 item in the queue) when you start this function, but by the time you get to it (after executing other items like timers or other subscriptions) one of two things could have happened:

- a new message could have arrived, replacing the one you took already (depth 1, keep last)
- the lifespan of the message could have expired, meaning the queue might have been empty by the time you got to it (had you not preemptively taken it)
- if there's a presentation order QoS in use (e.g. source timestamp rather than received order) a new message may also "jump ahead" in the queue rather than replacing a message

So, in short, you should try to minimize the time between taking data and actually executing it to avoid discrepancies like I described above.

Reply:
Correct me if I'm wrong, but after reading the references [paper](https://pdfs.semanticscholar.org/4a9e/b9a616c25fd0b4a4f7810924e73eee0e7515.pdf) it appears that the LET execution model assumes task runtime (ie. callbacks in this case) is bounded by some fixed activation duration. Otherwise, how can we get the promised determinism from executor?

Reply:
Yes, I see your point. We will document the use-cases for the static LET-executor and improve the implementation. We also see the issue of the lifespan. We are thinking about using a trigger which will then initiate rcl_take and processing the callbacks. In general, our use-case are domain specific, e.g. often we see sense-plan-act and cause-effect chains in general. Here the user would like to wait for an initial trigger (e.g. laser scan) and then process all sensor inputs at that time point. The let-executor is way to provide the user a simple API to specify such behavior. We will describe this in more detail in the pull request to rclc.